packages/itwinui-react/src/utils/hooks/useVirtualScroll.ts

```diff
@@ -0,0 +1,40 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { useVirtualizer } from '@tanstack/react-virtual';
+import type { Virtualizer } from '@tanstack/react-virtual';
+import { useLayoutEffect } from './useIsomorphicLayoutEffect.js';
+
+/**
+ * Wrapper over `useVirtualizer` from `@tanstack/react-virtual` that sets the index attribute to
+ * `data-iui-index` and adds a `useLayoutEffect` for scrolling to a provided index on render.
+ *
+ * @example
+ * const virtualizer = useVirtualScroll({
+ *  count: item.length,
+ *  getScrollElement: () => parentRef.current,
+ *  estimateSize: () => 30,
+ *  scrollToIndex
+ * });
+ */
+export const useVirtualScroll = (
+  params: Parameters<typeof useVirtualizer>[0] & { scrollToIndex?: number },
+): Virtualizer<Element, Element> => {
+  const { scrollToIndex, ...rest } = params;
+
+  const virtualizer = useVirtualizer({
+    indexAttribute: 'data-iui-index',
+    overscan: 10,
+    ...rest,
+  });
+  useLayoutEffect(() => {
+    setTimeout(() => {
+      if (scrollToIndex) {
+        virtualizer.scrollToIndex(scrollToIndex, { align: 'auto' });
+      }
+    });
+  }, [scrollToIndex, virtualizer]);
```

Comment:
Thinking about https://github.com/iTwin/iTwinUI/pull/2092#discussion_r1665813754, can we improve the way scrollToIndex is handled?

Instead of using a layout effect that runs when the passed index changes,we could expose a wrapper over `virtualizer.scrollToIndex` which adds the `setTimeout`, and adds the default `align` while allowing it to be overridden. This way, we could use `align: "auto"` in `ComboBox` and maybe `align: "center"` in `Table`.

```jsx
const scrollToIndex = React.useCallback((index, options) => {
  setTimeout(() => {
    virtualizer.scrollToIndex(index, { align: 'auto', ...options });
  });
}, [virtualizer.scrollToIndex]);

return React.useMemo(() => ({
  ...virtualizer,
  scrollToIndex,
}, [virtualizer, scrollToIndex]);
```

Reply:
Actually it looks like we don't even need to manually create `scrollToIndex`. We can pass a custom function to [`scrollToFn`](https://tanstack.com/virtual/latest/docs/api/virtualizer#scrolltofn), where we simply wrap the exported `elementScroll` with `setTimeout`.

Reply:
For some reason, passing in a `scrollToFn` that is the `elementScroll` wrapped with `setTimeout` didn't work. Exposing the wrapper around the `virtualizer.scrollToIndex` function does work, however. Is the plan to still call this function from within a layout effect, though? Is there some better event handler that we can call this function inside of?

Reply:
We can use `useLayoutEffect` (in `ComboBoxMenu` and `Table`) for now, but in the future I would like to try to remove the `useLayoutEffect` because it is [discouraged](https://react.dev/learn/you-might-not-need-an-effect#sharing-logic-between-event-handlers).

Reply:
Okay, I've changed the hook to pass back the wrapped function in this [commit](https://github.com/iTwin/iTwinUI/pull/2061/commits/a1adf575c050c874ec79d1f40a146118df99dad3) and updated the other PRs to account for the changes made.